dask/array/rechunk.py

```diff
@@ -246,6 +238,33 @@ def intersect_chunks(old_chunks, new_chunks):
     return cross
 
 
+def _validate_rechunk(old_chunks, new_chunks):
+    """Validates that rechunking an array from ``old_chunks`` to ``new_chunks``
+    is possible, raises an error if otherwise.
+
+    Notes
+    -----
+    This function expects ``old_chunks`` and ``new_chunks`` to have matching
+    dimensionality and will not raise an informative error if they don't.
+    """
+    assert len(old_chunks) == len(new_chunks)
+
+    old_shapes = tuple(map(sum, old_chunks))
+    new_shapes = tuple(map(sum, new_chunks))
+
+    for old_shape, old_dim, new_shape, new_dim in zip(
+        old_shapes, old_chunks, new_shapes, new_chunks
+    ):
+        if old_shape != new_shape:
+            if not (
+                math.isnan(old_shape) and math.isnan(new_shape)
+            ) or not np.array_equal(old_dim, new_dim, equal_nan=True):
```

Comment:
Using `np.array_equal` avoids issues with the provenance of nan values (https://github.com/dask/dask/pull/10157#discussion_r1203663937, https://github.com/dask/distributed/pull/7856/files#r1205199004)

Reply:
Makes sense, I think.